File: recipes/hostui/gudule-greeting-hostui-flutter/lib/src/client/greeting_client.dart
import 'package:grpc/grpc.dart';

import '../../gen/greeting/v1/greeting.pbgrpc.dart';
import 'daemon_launcher.dart';
import 'greeting_target.dart';

class GreetingClient {
  ClientChannel? _channel;
  GreetingServiceClient? _stub;
  final DaemonLauncher _launcher;

  GreetingClient({DaemonLauncher? launcher})
      : _launcher = launcher ?? DaemonLauncher();

  bool get isConnected => _stub != null;

  Future<void> connect(GreetingEndpoint endpoint) async {
    await close();
    final channel = await _launcher.start(endpoint);
    _channel = channel;
    _stub = GreetingServiceClient(channel);
  }

  Future<ListLanguagesResponse> listLanguages() async {
    return _stub!.listLanguages(ListLanguagesRequest());
  }

  Future<SayHelloResponse> sayHello(String name, String langCode) async {
    return _stub!.sayHello(
      SayHelloRequest(name: name, langCode: langCode),
    );
  }

  Future<void> close() async {
    final channel = _channel;
    _stub = null;
    _channel = null;
    await _launcher.stop(channel);
  }
}

File: recipes/hostui/gudule-greeting-hostui-kotlin/src/main/kotlin/greeting/gokotlin/grpc/GreetingClient.kt
package greeting.gokotlin.grpc

import greeting.v1.Greeting
import greeting.v1.GreetingServiceGrpcKt
import io.grpc.ManagedChannel
import kotlinx.coroutines.runBlocking
import org.organicprogramming.holons.Connect

data class GreetingLanguage(
    val code: String,
    val name: String,
    val native: String,
)

class GreetingClient(
    private val channel: ManagedChannel,
) {
    private val stub = GreetingServiceGrpcKt.GreetingServiceCoroutineStub(channel)

    suspend fun listLanguages(): List<GreetingLanguage> {
        return stub.listLanguages(Greeting.ListLanguagesRequest.getDefaultInstance()).languagesList.map { language ->
            GreetingLanguage(
                code = language.code,
                name = language.name,
                native = language.native,
            )
        }
    }

    suspend fun sayHello(name: String, langCode: String): String {
        val request = Greeting.SayHelloRequest.newBuilder()
            .setName(name)
            .setLangCode(langCode)
            .build()
        return stub.sayHello(request).greeting
    }

    fun close() {
        runBlocking { Connect.disconnect(channel) }
    }
}
